concert/ext/tangoservers/camera.py

```diff
@@ -0,0 +1,240 @@
+"""
+Tango devices for dummy concert cameras sending images over zmq sockets.
+"""
+import asyncio
+import numpy as np
+import tango
+from concert.base import StateValue
+from concert.devices.cameras.dummy import Base, Camera, FileCamera
+from concert.networking.base import ZmqSender
+from tango import DebugIt, InfoIt
+from tango.server import attribute, AttrWriteType, command, Device, DeviceMeta
+
+
+def _run_asyncio(func, *args, **kwargs):
+    # This will break the second the owner of the *func* needs a specific event loop
+    loop = asyncio.new_event_loop()
+    try:
+        asyncio.set_event_loop(loop)
+        return loop.run_until_complete(func(*args, **kwargs))
+    finally:
+        loop.close()
+
+
+class TangoCamera(Device, metaclass=DeviceMeta):
+    """
+    Tango device for the dummy concert camera sending images over a zmq socket.
+    """
+
+    endpoint = attribute(
+        label="Endpoint",
+        dtype=str,
+        access=AttrWriteType.READ_WRITE,
+        fget="get_endpoint",
+        fset="set_endpoint"
+    )
+
+    def __init__(self, cl, name):
+        self._sender = ZmqSender()
+        self._params_initialized = False
+        super().__init__(cl, name)
+
+    async def init_device(self):
+        self.info_stream('%s init_device', self.__class__.__name__)
+        await super().init_device()
+        self._index = 0
+        if not self._params_initialized:
+            camera = await Base()
+            for param in camera:
+                self.debug_stream("Adding `%s' attribute", param.name)
+                default = await param.get()
+                write_type = AttrWriteType.READ_WRITE if param.writable else AttrWriteType.READ
+                has_unit = hasattr(param, 'unit')
+                if has_unit:
+                    default = default.magnitude
+                if not isinstance(param, StateValue):
+                    attr = tango.Attr(
+                        param.name,
+                        tango.utils.TO_TANGO_TYPE[type(default)],
+                        write_type
+                    )
+                    self.add_attribute(
+                        attr,
+                        r_meth=self.dynamic_getter,
+                        w_meth=self.dynamic_setter if param.writable else None
+                    )
+            self._params_initialized = True
+
+    def dynamic_getter(self, attr):
+        # If this were async def Tango would complain about it never being awaited
```

Comment:
This is one of the reasons I will not change the format strings + eval in the reco server right now.